### PR TITLE
Add regexes to ignore entries and evaluate special cases

### PR DESCRIPTION
**Pull request by AlexanderLill**

### CODE REVIEW

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:48]: Maybe move the if statements to its own methods (see is_fee). This is mostly to reduce the complexity of the overall function.
This could look something like:
```suggestion
        elif self.is_special_entry_interest(booking_type):
            category = "Zinsen"
        elif self.is_special_entry_fee(booking_type):
            category = "Gebühren"
        elif self.is_ignorable(booking_type):
            category = "Ignored"

    def is_ignorable(self, booking_type):
        return self._config.get_ignorable_entry_regex() and self._config.get_ignorable_entry_regex().match(booking_type)

    def is_special_entry_interest(self, booking_type):
        return self._config.get_special_entry_regex() and self._config.get_special_entry_regex().match(booking_type) and self.value() >= 0

    def is_special_entry_fee(self, booking_type):
            return self._config.get_special_entry_regex() and self._config.get_special_entry_regex().match(booking_type) and self.value() < 0
```

Overall (in retrospect) not too happy with how much this if branching explodes :) I should have come up with a better way to handle these different categories.

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:59]: Why is this required? Are there statements without date? not so nice. hmm.

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:65]: Hm. would this be cleaner?
```
        if self._statement[self._config.get_booking_date()]:
            statement_date = datetime.strptime(
                self._statement[self._config.get_booking_date()], self._config.get_booking_date_format()
            ).date()
        else:
            statement_date = datetime(1970, 1, 1).date()
        return statement_date
```

Could you also add a sample entry to the mintos test data to verify that an empty entry is correctly processed? :)

**Review by ChrisRBe — CHANGES_REQUESTED** [review]
Let me know your thoughts on my comments :)

**Comment by AlexanderLill** [comment]
@ChrisRBe thanks for the very useful feedback :) I might be able to work on it on the weekend and also add some tests. I saw the no-date entry the first time, but maybe it was some kind of incomplete transaction on mintos' side, it was the last row in my csv export.

File: src/p2p_account_statement_parser.py
Comment (AlexanderLill) [reply src/p2p_account_statement_parser.py]: done, thanks :)

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: Refactored so that does not show up anymore. I knew, but sometimes I tend to be overly explicit with these if statements ^^

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: Refactored so that does not show up anymore, but I agree, it was probably not necessary

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: Thanks, I tried to make it a little more clear in the refactored version, let me know what you think :)

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: I had a downloaded csv with a line without date, but in the file I downloaded today all entries have a date.

So we could keep it to be more resilient for missing dates, as otherwise the parser just throws an error and stops.

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: Yes that looks a lot better, thanks!

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py]: I also put the code for handling the special mintos case in a separate function, I am not too happy about the whole part, but the format of the csv does not make it easy to parse the category :|

**Comment by AlexanderLill** [comment]
From my side I would now wait for further feedback. Unfortunately the current version does not pass the codeclimate requirement for complexity 5 (it has complexity 6).

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:44]: I like this approach! I have been struggling to find a better/different way from what I implemented first.

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:49]:
```suggestion
        if not category:
```

empty string is handled as Boolean False :)

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:57]:
```suggestion
        """
        This is currently a special case for the Mintos "discount/premium" secondary market transactions parsing,
        where an entry might be a fee or an income depending on its sign.
        
        :param value: how much money was returned/paid
        
        :return: Zinsen if value >= 0; Gebühren if < 0; Ignored in any other case
        """
```

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:64]: Please correct me if I'm wrong, the else statement is only for the case where value is not a number, right? hm. Should work I guess.

**Review by ChrisRBe — CHANGES_REQUESTED** [review]

**Comment by ChrisRBe** [comment]
Hey @AlexanderLill, looks pretty good already. only one or two minor comments. Let me have a look at codeclimate, what he is bitching about.

**Comment by ChrisRBe** [comment]
> Hey @AlexanderLill, looks pretty good already. only one or two minor comments. Let me have a look at codeclimate, what he is bitching about.

I had another look and added a PR on your fork repository. Let me know what you think.

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py:49]: Will make sure to remove all these comparisons with "" :)

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py:64]: Yes exactly, it might be unexpected, but at least that way the "Undecided" will not make it into the next higher level where it is not handled.

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py:57]: That looks better, thanks :)

**Comment by AlexanderLill** [comment]
@ChrisRBe thanks for the feedback, I merged your branch into mine and implemented your feedback. For me this PR on github is getting a bit confusing (especially with the comments on code parts that have changed in the meantime) but I hope I did not forget anything :)

**Comment by ChrisRBe** [comment]
Okay. There's only 2 issues remaining on coding content regarding duplicate code in the writer. I accepted those already and really should have a look at later.

**Comment by ChrisRBe** [comment]
> @ChrisRBe thanks for the feedback, I merged your branch into mine and implemented your feedback. For me this PR on github is getting a bit confusing (especially with the comments on code parts that have changed in the meantime) but I hope I did not forget anything :)

I will have another look, but it looks really good already :)

File: src/Statement.py
Comment (ChrisRBe) [inline src/Statement.py:119]: Could you please add some test data in src/test/testdata/mintos.csv and update the unit test so that we verify the correct behaviour? :) That would be fantastic!

**Review by ChrisRBe — CHANGES_REQUESTED** [review]

File: src/Statement.py
Comment (AlexanderLill) [reply src/Statement.py:119]: Done :) But in my portfolio I only have cases for "Gebühren" so I am not 100% sure on the other if-branch.

Comment (ChrisRBe) [reply src/Statement.py:119]: one fix outstanding for the black findings :) will have a look after I finish gaming.

Comment (ChrisRBe) [reply src/Statement.py:119]: You mentioned that you have one file without any date added? Could you add that as well to the test data please? Just to validate also that side of the branch :)
And hm.. looks like the if and elif branches are covered... at least according to my local checks. Missing would be the last one. Any idea? Maybe set the value field to a string for testing?

Comment (AlexanderLill) [reply src/Statement.py]: Done :) In fact the `else` branch is never executed, as casting something not a number to a float before (in `get_value`) fails anyway, so we don't need that case here

**Review by ChrisRBe — APPROVED** [review]

**Comment by ChrisRBe** [comment]
@AlexanderLill looks good overall now :) Let me know when you want to merge the PR. 👍

**Comment by AlexanderLill** [comment]
@ChrisRBe great, from my side it is ready to merge, feel free to do it whenever you can :)

**Pull request merged**
